### DISCUSSION TITLE
The absence of records in Broker's data model

### DISCUSSION THREAD

**Discussion by ckreibich** [zeek/zeek#2879] [Ideas]
Recent discussions around backward-compatibility of changes to records flagged how particularly exposed users of Broker's Python bindings are to such changes. Mapping Zeek's records to [Broker's data model](https://docs.zeek.org/projects/broker/en/master/data.html) expects records as a tuple of values, with each member in its own Broker-model rendering. The vector must cover the full list of fields, including `None`s for optional ones. As a result, any change to the record definition means you must adapt your Python implementation to reflect it. This is true for both the traditional wire format and the new WebSocket one.

I am wondering whether we should extend Broker's data model to allow record types, with a serialization as a key/val store.

Does anyone recall the historical reasoning for stopping short of explicit record support in the model?

@J-Gras, @Chukudubem, @Neverlord, @rsmmr, fyi

**Comment by vpax** [comment]
I'm wondering about efficiency tradeoffs, and in particular whether there's a way to support an initial phase that agrees on the expected positions, plus support for gaps, such that the common case of sending around a large number of agreed-upon records can proceed with no need to do per-record field lookups.

**Reply by ckreibich** [reply]
Agreed, the Python use case is certainly the exception. This might be as simple as serializers/deserializers supporting both the existing vector-based and an alternative key/val-based rendering.

**Reply by Neverlord** [reply]
> way to support an initial phase that agrees on the expected positions

That could work with a tightly controlled Zeek cluster, but I don't think we can get that to work with clients that dynamically appear at runtime. In particular with the WebSocket interface, clients may appear and disappear at any time. That would render any data useless if they somehow had to be present at the time of setup to learn the schema.

> Agreed, the Python use case is certainly the exception. This might be as simple as serializers/deserializers supporting both the existing vector-based and an alternative key/val-based rendering.

I don't quite get that idea. Where would these serializers/deserializer be located? At Zeek? Zeek has no knowledge about whether or not 3rd-party clients might appear at some time. At Broker? At least until now, Broker has no knowledge about the kind of data that travels through the topics. I think if we do that at the Broker side, it would become both expensive and brittle. Expensive, because we need to look into every message that crosses over a WebSocket and unbox / reassemble nested data structures on the fly. Brittle, because there is no definitive way to know what is a Zeek record. We can try to guess, but I think it's just asking for trouble.

If we have data that packing record types as vector is a must for performance (did we ever do an evaluation on that?), than I think we would need some sort registry to put the schemas. That way, a client that appears may get the meta data at least on demand. But of course there's the question where to put that registry. Self-contained data is just much, much more easy to work with and more robust.

A point that seems missing so far is that clients connected over WebSocket may not just consume data but also generate it. I don't know whether the Zeek Agents already send records to Zeek, but I think it's something that is worth opening up. As long as we only ever *add* fields to a record type, that should be fine. Clients that don't know the new fields can just ignore them. Missing fields are considered null/none or later added fields must provide some default.

> I am wondering whether we should extend Broker's data model to allow record types

I think a dedicated "record" data type in Broker would make things simpler. Rendering them as table adds overhead (because a key can be any data type, not just strings) and of course ambiguity. A dedicated record type can take some shortcuts (like only allowing strings as keys).

**Reply by Neverlord** [reply]
> If we have data that packing record types as vector is a must for performance (did we ever do an evaluation on that?)

After reading again, I think this sentence sounds a bit off/harsh. Just to clarify: I didn't mean it that way. 🙂

Just genuine curiosity whether anyone remembers some kind of performance evaluation on that or even better still has some data from back then. The data model of Broker was already in place when I joined. If we just don't have any data, that's fine. I'm just advocating to have evaluation/data before ruling out solutions.

**Reply by ckreibich** [reply]
It's all good Dominik! I think the performance question is interesting, but I don't really mean to change record rendering everywhere. I'm just looking to loosen the strictness of rendering Zeek records.

Let's say we don't change Broker at all. We could change Zeek so that it supports records not just in form of Broker vectors, but alternatively also as Broker tables. Nothing changes about Zeek clusters or performance because it all keeps running as-is. But for use cases like the Python bindings, where some freedom in the rendering is desirable, it would help:

When receiving a table-based record, the Python side could just grab the fields it cares about. If a field got dropped, it's not there, which can be handled better than a vector with changing lengths or member meanings.

When sending a table-based record, the Python side doesn't need to know the full set of fields. Zeek may determine that the received data is insufficient to instantiate a record because fields are missing, but that too can be handled robustly.

One catch is that in order for the Python end to receive table-based records, the other side would need to learn somehow that it should use tables. It would be nice if as part of the connection set-up the endpoints could send some config info — a config table, say. That would be nice for other reasons too:  for example, it'd spare the Cluster framework from building up a table of Broker IDs to node names — the endpoint could just send that along at the very beginning.

Does that help?

**Reply by awelzel** [reply]
> One catch is that in order for the Python end to receive table-based records, the other side would need to learn somehow that it should use tables. It would be nice if as part of the connection set-up the endpoints could send some config info

(considering websocket part and external client usage of this only)

IMO "table-based records" should be/have been the default. Maybe on a `/v2`. It could also be opt-in via query parameter in the URI or HTTP header in the initial request on `/v1`. Possibly could also be done backwards compatible by stuffing `@field-name` elements into the elements of a "record vector" so it looks like below:
```
      "@data-type": "vector",
          "data": [
            {
              "@field-name": "x",
              "@data-type": "count",
              "data": 42
            },
            {
              "@field-name": "y",
              "@data-type": "count",
              "data": 4711
            }
          ]
```

It's then implicit that vectors with `@field-names` are actually records, but knowledgeable clients can do things better upon receiving these. The explicit "record" type might just be as follows (and somewhat closer/usual to how a HTTP/JSON API would be).
```
      "@data-type": "record",
          "data": {
            "x": {
              "@data-type": "count",
              "data": 42
            },
            "y": {
              "@data-type": "count",
              "data": 4711
            }
          }
```

Many systems seem fine consuming conn.log, ssl.log, ... in JSON format, so not sure about the performance/optimization. Specifically for the websocket data representation: it's already verbose JSON and that record optimization seems to actually hinder broader/easier integration and adoption.

**Reply by ckreibich** [reply]
> (considering websocket part and external client usage of this only)

I think that's actually right on the mark — I like the idea of focusing this on the WebSocket transport. It's where we're going for the Python bindings anyway, users of this transport are arguably already in the "comfort, not performance" class, and via the WebSocket framing we get nice versioning support. If you have a Zeek-external use case that needs an efficient transport, it seems reasonable to me to require linking with Broker.

My understanding of Dubem's use case and Robin's zeek-agent-v2 is that event rates wouldn't be a concern, but holler if I'm off there.

**Reply by ckreibich** [reply]
> via the WebSocket framing we get nice versioning support

Oops, that confused Broker vs. Zeek's use of it. We'd want some way to convey the Zeek-level dialect the endpoint speaks, so I'm back to wanting some way to express configurability at the start.

>  stuffing @field-name elements

Could do, yeah, though all in all an explicit switch to a table seems nicer.

**Reply by Neverlord** [reply]
We probably need to define a mapping to `broker::table` in any case. That "trick" with adding extra annotations to a `vector` won't work there. The native format is also versioned, but we would need to have a mapping from v1 to v2 to make sure we can mix v1 and v2 endpoints in the network. Otherwise, we might lose information. On the /v1 WebSocket endpoints, the records would then also show up as a table.

The table representation of a record could just have `__META-TYPE__` set to `__RECORD__` (`__` prefix+suffix is just an idea to have some sort of reserved "namespace" for such things).

On `v2`, these could then get flattened into the new, more efficient record data type. If we go down that route, I think we should really think through the types in Broker to see if we have actual good representation for all Zeek types or if we could use some extra types. I'd rather add a couple more types with v2 than having to do a v3 soon after.

Or, we just use the tables with some special "marker" field and see whether that's enough? May not be pretty, but we could still compress them at some point in the future with a dedicated type if performance turns out to be an issue with that representation. That would then boil down to a Zeek-side change only for now. For the performance impact, we could build a simple benchmark that renders some common record types in Zeek in both versions and then sees how many of these we can push through Broker per second.

Of course it's also an option to say the native wire-format v2 is incompatible with v1. However, we have to have the mapping between v1 and v2 in any case if we still want to offer the v1 API to WebSocket clients. Would feel weird to omit that mapping for the native format.

**Comment by rsmmr** [comment]
> Does anyone recall the historical reasoning for stopping short of explicit record support in the model?

Yeah, performance indeed I think, maybe also ease of implementation. 

(Also, the other notable piece missing is typing, but that applies to Broker values generally.)

**Comment by ckreibich** [comment]
Okay ... I recently learned a few things about Broker. Might well be old news for most, but here's where I land as a result:

- We are discussing two layers — Broker with its native format & WebSockets, and Zeek's internal mapping of its types onto Broker's data model (as per `src/broker/Data.cc`). The latter isn't versioned, based on the assumption that in a Zeek cluster everything runs the same Zeek version.

- There's no explicit typing — there's always going to be some node-local assumption of what a given type looks like as it's read/written, with the potential of other nodes disagreeing. (That's unless we sent type descriptions etc — I'll ignore this for the moment.)

- There's (obviously) Zeek-level event handling for communication purposes: a received event gets explicitly forwarded from one topic to another, possibly repackaged, perhaps responded to on another topic, RPC-style, etc — all ad-hoc. (One of the "Big Ideas" items considers formalizing these communication models.)

- There's also Broker-level message handling: Broker supports internal message relay across nodes on a given topic, all below the Zeek layer. We don't use this in Zeek: the script-level `Broker::forward_messages = F` disables it. This was the gist of gateways and relay of messages for zeek-agent we discussed in the past.

- Ignoring for the moment the fact that it's disabled, this property means that the notion of one node (a Python client) telling a neighbor (a Zeek cluster node) at the Zeek level that it'd like to receive records in a certain serialization format doesn't fit: the creator of an event might be somewhere else in the cluster, with no idea of the local preference. So in order to express a potential serialization preference from one node to its neighbor, we'd have to focus on the Broker level, which feels pretty bad — that's building more and more magic into Broker.

- Seems there's agreement that the cleanest way forward would be for records to be rendered as Broker tables (or a new explicit record type), which implies performance questions.

I don't really like the idea of tweaking the current Broker vectors to somehow better accommodate Zeek's records — Broker's data model should fit naturally. At least, I don't think we should do this next. My suggestion is to actually examine the impact of a switch to table-based records in Zeek. This doesn't strike me as that much work: it looks like a rewrite of a few components of `src/broker/Data.{cc,h}`, plus some benchmarking (say via os-perf or custom benchmark code). It's interesting and informs next steps.

I also don't think this is especially urgent given how much else is going on right now. The situation for Python / non-Zeek endpoints isn't great, and that alone was worth highlighting. But it's also not a showstopper.

Thoughts welcome...

**Comment by bbannier** [comment]
Just to add another user data point, I recently wrote a serializer/deserializer for records and hit no real issues with the current vector-based approach.

```rust
// Define a type. This declares the expected shape of the data, the `#[derive(ZeekType)]`
// causes emiting extra code for serializing/deserializing.
#[derive(ZeekType)]
struct Record {
    a: i64,
    b: u64,
}

// User-level value.
let r = Record { a: -32, b: 1024 };

// Can serialize to Zeek WebSocket wire format.
let value = Value::from(r);
assert_eq!(
    value,
    Value::Vector(vec![Value::Integer(-32), Value::Count(1024)]));

// Can deserialize from Zeek WebSocket wire format.
let r = Record::try_from(value).unwrap();
assert_eq!(r, Record { a: -32, b: 1024 });
```

As an external data point, Protobuf [does not serialize field names either](https://protobuf.dev/programming-guides/encoding/) and just relies on order.

If we assume a world where clients bake in some form of a schema snapshot (e.g., by them just describing the shape of the expected data like I did above), working with different versions of the schema requires some pretty simple rules which are overall very similar to [what Protobuf recommends](https://protobuf.dev/programming-guides/field_presence/). We seem to already follow these to some extent, but being strict would end up being some work (with Zeek any event could be an external API).

To not cause an API break:

- As a provider of some record schema (probably: provider of some shared event and its data types) the schema for a record currently must follow some rules:
  - record fields must never be reordered
  - all non-`&optional` fields must be serialized
  - in a stable schema any new field must be appended to the end of the record and must be marked `&optional` to keep users with older schema versions working
  - fields must never be removed
- A client should also follow some rules:
  -  `&optional` fields must not be required, e.g., in Python they'd be modelled as `Optional[T]`
  - since Zeek remotes might add new fields, any data in a vector record after the last known field should be silently ignored (the remote might have added new `&optional` fields per third point above)
- With above rules encoding a tail of `None` fields is not really needed (clients would know to set missing values as `None` anyway to work with different schema versions) and could be dropped. In religiously evolved Protobuf schemas there will often be many `optional` fields from previous versions, so not having to serialize long tails can save space in the serialized messages. 

Not getting serialized field names from Zeek was just fine as in the deserializer into my language's record type I only cared about the actual data. Since I was working in a statically compiled language I wouldn't have been able to dynamically create fields in deserialized record types anyway. Even a dynamic language like Python could be made to work like this with frameworks like [pydantic](https://docs.pydantic.dev/latest/). In fact not requiring the names allowed me to name fields any way I liked.

Switching from serializing as vectors to tables doesn't make most of this go away. AFAICT it would only allow the following:

- Fields can now be reordered.
- `&optional` fields can now appear anyway.
- `&optional` fields can now be removed
- Using tables would effectively be distributing of named bags of data, but to interpret that bag the name would suddenly carry a lot of semantics. This seems very powerful, but to make this work well we'd need to introduce a new schema describing semantics of field names; I am not sure I am looking forward to that 😅.

Using a table-based serialization introduces new requirements:

- one might need to remember the wire-format name of a field, e.g., `u64` is a keyword in my target language so if I renamed such a field I would have to remember its wire-format name for serialization.

With records as vectors we could during an API break:

- make `&optional` fields required (making a field `&optional` is not an API break as long as we keep sending its data) 

All that said, like mentioned by @Neverlord above, the current format of tables is _very wide_ since they allow any value type as key, and I had trouble exposing Zeek WebSocket tables as native tables in my target language. There floats (i.e., `real` values from Zeek) cannot be compared for equality so they cannot be used as table keys, and I needed to expose tables as vectors of pairs. There might be a case for introducing more narrowly defined tables, but I am unsure it is related to records for me.

**Reply by bbannier** [reply]
I guess another way to put this: as a user the current wire format is not inconvenient, and a discussion about a different schema is probably mostly driven by the constraints it introduces on how event providers can evolve their schemas.

EDIT:

It seems the major issue with records as vectors is that it is impossible the ever remove an `&optional` field while still giving clients a way to migrate across the API break. For use of records in scripts we can mark a field as `&optional` and mark it deprecated, and can remove them during API break. Script against that type not using the deprecated field keep working across the break. This seems impossible or at least very hard for WebSocket clients as they cannot write code which works both before and after the break.

Protobuf solves this by numbering all fields, e.g.,

```proto

message SearchRequest {
  string query = 1;
  int32 page_number = 2;
  int32 results_per_page = 3;
}
```

This allows dropping optional fields again if one never reuses their field number. They also use a wire format which compresses repeated fields, so something like "42 optional but unset fields here" is cheap to encode. I think Zeek script has no equivalent feature.

I guess I have just convinced myself that a dictionary/table-based encoding would allow evolving the schema of record types in a better way than the current vector-based one.
